tree/tree/src/TEventList.cxx

```diff
@@ -368,7 +376,11 @@ void TEventList::SetName(const char *name)
    //  We must update the hashlist if we change the name
    if (fDirectory) fDirectory->Remove(this);
    fName = name;
-   if (fDirectory) fDirectory->Append(this);
+   if(fDirectory && fDirectory->GetAcceptsRegisters()){
```

Comment:
That seems to be a very odd semantic!
```
TEventList *elist = .....
// do stuff so that elist->GetDirectory() != nullptr
// and also so that (indirectly)
elist->GetDirectory()->SetAcceptRegisters( false );
...
// and somewhere else
assert( elist->GetDirectory() );

elist->SetName( newname );

// and now 'shockingly' we have
assert( elist->GetDirectory() == nullptr );
```

Reply:
It's really the same semantyc as what is currently written in TH1.cxx
```cpp
   if (TH1::AddDirectoryStatus()) {
      fDirectory = gDirectory;
      if (fDirectory) {
         fFunctions->UseRWLock();
         fDirectory->Append(this,kTRUE);
      }
   }
   ```   // Insert ourself into the current directory.
   // FIXME: This is very annoying behaviour, we should
   //        be able to choose to not do this like we
   //        can with a histogram.
   if (fDirectory) fDirectory->Append(this);
```

Just that instead of a static flag that's useful only for histograms, there is a flag specific to the file the object is trying to register with. This should allow for a more flexible ownership management, at least in principle. I'm not saying this is the right implementation, but what about the generic idea? This is for example hinted at in https://github.com/root-project/root/blob/9a019c901d8d26d5c1aa389d5c1818d1e3be3b96/tree/tree/src/TTree.cxx#L899-L903

Reply:
Maybe your comment above was just for the usage of this pattern in a `SetName` function. Yeah maybe that's wrong, I just grepped for usages of `fDirectory->Append` in the repository and changed all of them.